nshlib/nsh_parse.c

```diff
@@ -115,6 +129,14 @@ struct nsh_memlist_s
 };
 #endif
 
+#ifdef CONFIG_NSH_ALIAS
+struct nsh_alist_s
+{
+  int nallocs;
+  FAR struct nsh_alias_s *allocs[CONFIG_NSH_ALIAS_MAX_AMOUNT];
```

Comment:
why need another table to keep the expanded variable if nsh_alias_s already has exp field? I think it's more simple to iterate all alias and clear exp directly.

Reply:
> It's more simple to iterate all alias and clear exp directly after the end of expansion.
I thought of this and decided to make the list anyway. 

It is there so you don't need to traverse the whole alias list after every command. I for example have >50 aliases in my use case. In the normal case when using the terminal no alias expansion is performed.

Example:
```
$ alias foo1=bar
$ alias foo2=bar
...
$ alias foo50=bar
$ ls
```
Here after executing the ls command, even though no alias expansion has been performed you would have to go through all 50 aliases to see if they have been expanded. Instead now we have a "expanded aliases list" that is empty and nothing needs to be done.

There is a purpose for the expanded alias list, the exp flag and the rem flag.
1. The expanded alias list contains every expanded alias, so they can be cleared and if necessary, removed from the alias list after a command has been executed
2. The exp flag is already explained above, an alias shall be expanded only once
3. Rem field is also needed as the expanded alias string is not copied onto the parsed command line, simply a reference is returned and this means the alias data must be retained until the command has been executed. 

When is the rem field relevant ? Well, some insane corner cases can appear...

```
$ alias ua='unalias -a'
$ ua
```
Here the string behind the alias removes all aliases, but it is not safe to remove ua before the command unalias -a has been executed.

Reply:
If so, look like it's better to manage the expansion list directly:
https://github.com/apache/nuttx-apps/pull/1654/files#diff-7ba973a1f25993d049f08f4b5d54ac41f9a0955f4f2642c0ddb069dde7c4c5e4R124-R130
for example, adding an expansion and move the alias from alist to expansion list vice vesa.

Reply:
I thought of this as well, but it has side effects too.

1. The ordering of the aliases changes when pumping from one list to another. Although the significance of this is mostly cosmetic.
2. The corner case above would not work any more, as executing `$ ua ` would temporarily remove ua from the valid alias list (moving it to the expanded list). So all aliases *besides* ua would be removed.

Another example:

```
$ alias foo='ls;unalias foo'
$ foo
```
This would fail to perform the "unalias foo" part, as it has been (temporarily) plucked from the list of known aliases (it cannot be found anymore).

I know these corner cases are insane usages of aliases, but the standard does not prohibit making such aliases.

I did explore other options for the separate NSH_ALIASLIST storage but all of them had some corner case that did not work. This implementation does at least work.

Reply:
Ok.